e2e_tests/tests/cluster/test_ship_logs.py

```diff
@@ -0,0 +1,512 @@
+        # Most dev machines probably will be fine with small timeouts, but CI machines might be
+        # slower and we allow up to 0.2 seconds of slop.
+        timeouts = [0.001, 0.2]
+        for timeout in timeouts:
+            cmd = mkcmd(
+                f"""
+                import sys
+                import time
+                print("1", file=sys.stdout, flush=True)
+                time.sleep({timeout})
+                print("2", file=sys.stderr, flush=True)
+                time.sleep({timeout})
+                print("3", file=sys.stdout, flush=True)
+                time.sleep({timeout})
+                print("4", file=sys.stderr, flush=True)
+                time.sleep({timeout})
+                print("5", file=sys.stdout, flush=True)
+                time.sleep({timeout})
+                print("6", file=sys.stderr, flush=True)
+                """
+            )
+            with ShipLogServer() as srv:
+                exit_code = self.run_ship_logs(srv.master_url(), cmd)
+            assert exit_code == 0, exit_code
+            if "".join(srv.logs) == "1\n2\n3\n4\n5\n6\n":
+                # Success
+                break
+            elif timeout == timeouts[-1]:
+                # Failed, even on the highest timeout
+                raise ValueError("".join(srv.logs))
+
+    @pytest.mark.e2e_cpu
+    def test_signal_forwarding(self) -> None:
+        cmd = mkcmd(
+            """
+            import signal
+            import time
+
+            def handle_sigint(*arg):
+                print("caught sigint!")
+
+            signal.signal(signal.SIGINT, handle_sigint)
+
+            print("ready!", flush=True)
+
+            time.sleep(5)
+            """
+        )
+        with ShipLogServer() as srv:
+            # Start a subprocess so we can signal it.
+            env = {
+                "DET_MASTER": srv.master_url(),
+                "DET_SESSION_TOKEN": "token",
+                "DET_TASK_LOGGING_METADATA": "{}",
+                "DET_SHIPPER_EMIT_STDOUT_LOGS": "1",
+            }
+            fullcmd = [sys.executable, "-u", ship_logs.__file__] + cmd
+            p = subprocess.Popen(fullcmd, env=env, stdout=subprocess.PIPE)
+            assert p.stdout
+            try:
+                # Wait for the granchild log to indicate signals are set up.
+                for line in p.stdout:
+                    if b"ready!" in line:
+                        break
+                # Send a signal that is caught and logged, to test signal forwarding.
+                p.send_signal(signal.SIGINT)
+                for line in p.stdout:
+                    if b"caught sigint!" in line:
+                        break
+                # Send a signal that is not caught, to test for signal exit codes.
+                p.send_signal(signal.SIGTERM)
+                exit_code = p.wait()
+            finally:
+                p.kill()
+                p.wait()
+        assert exit_code == 128 + signal.SIGTERM, exit_code
```

Comment:
General "nice to have" to ask that assertions that test what these tests are designed to cover have a useful message. I'm not sure these provide an oncall value beyond just the bare assertion itself.

Not worth gating merging this PR over.

Reply:
Actually I might go strip these into being bare asserts.  The pytest magic is really good around bare asserts.  The reason this isn't bare is because the first draft of the test code didn't run in pytest and so didn't have the magic.

Reply:
I think that's an improvement, too.

Reply:
Nah, the pytest magic works fine as is, I'd rather just leave these alone honestly